src/actors/symbolication.rs

```diff
@@ -156,83 +159,87 @@ impl SymbolicationActor {
-                                    Err(inner)
-                                } else {
-                                    Ok(SymbolicationResponse::Pending {
-                                        request_id,
-                                        // XXX(markus): Probably need a better estimation at some
-                                        // point.
-                                        retry_after: 30,
-                                    })
-                                }
-                            }
-                        }
-                    })),
-            )
+        if let Some(timeout) = timeout.map(Duration::from_secs) {
+            Box::new(tokio::timer::Timeout::new(rv, timeout).then(move |result| {
+                match result {
+                    Ok((finished_at, response)) => {
+                        metric!(timer("requests.response_idling") = finished_at.elapsed());
+                        Ok(response)
+                    }
+                    Err(timeout_error) => match timeout_error.into_inner() {
+                        Some(error) => Err(error),
+                        None => Ok(SymbolicationResponse::Pending {
+                            request_id,
+                            // XXX(markus): Probably need a better estimation at some
+                            // point.
+                            retry_after: 30,
+                        }),
+                    },
+                }
+            }))
         } else {
-            Either::B(rv.then(clone!(requests, |result| {
-                requests.write().remove(&request_id);
-                let (finished_at, x) = result?;
+            Box::new(rv.then(move |result| {
+                let (finished_at, response) = result?;
                 metric!(timer("requests.response_idling") = finished_at.elapsed());
-                Ok(x)
-            })))
+                Ok(response)
+            }))
         }
     }
 
-    fn create_symbolication_request<F, R>(&self, f: F) -> Result<RequestId, SymbolicationError>
+    fn create_symbolication_request<F, R>(&self, f: F) -> RequestId
     where
         F: FnOnce() -> R,
         R: Future<Item = CompletedSymbolicationResponse, Error = SymbolicationError> + 'static,
     {
-        let request_id = loop {
-            let request_id = RequestId::new(uuid::Uuid::new_v4());
-            if !self.requests.read().contains_key(&request_id) {
-                break request_id;
-            }
-        };
+        let (sender, receiver) = oneshot::channel();
 
-        let (tx, rx) = oneshot::channel();
+        let requests = self.requests.clone();
+        let request_id = RequestId::new(uuid::Uuid::new_v4());
 
-        self.requests.write().insert(request_id, rx.shared());
+        // Assume that there are no UUID4 collisions in practice.
+        let evicted = requests.lock().insert(request_id, receiver.shared());
+        debug_assert!(evicted.is_none());
```

Comment:
We don't yet have any `debug_assert!()` calls.  In Python I've learned to avoid these constructs (`if __debug__`) because this almost always turns out to not work well.  Either the invariant holds and things are fine, or we silently have errors in production.  And in this case collisions are much more likely to occur in production than when testing so this will likely never do much good.

Reply:
Fair point. I'm making the following distinction here:
- Be defensive in production. The case will still be handled correctly, so I would not want to error out.
- Be strict in testing, to detect regressions early on.

WDYT?

Reply:
Isn't swallowing this silently what we do in production now?

If I understand correctly, if you happen to have evicted a request (so unlikely that this is all highly theoretical) then when the future of that evicted request gets dropped this request will be removed from the set of requests.  This means anyone polling for it will get an error saying this request doesn't exist.  Both the original request and this request will have its receiver dropped and sends on them will fail.  So I think when a collision happens both the affected requests are lost.

I'm fine with accepting this, it is extremely unlikely.  I'm still not sure panicing in debug mode does much to improve the situation though.

Reply:
I'm also not objecting to this btw, I did approve the PR after all :)

Reply:
While on a short walk I also realised that another way of looking at this is that if it is assertable under debug, then it could be written as a unit test.

Reply:
You convinced me, I'll remove them :)